README.md
# lax.js

Simple & light weight (<3kb gzipped) vanilla javascript plugin to create *smooth* & beautiful animations when you scrolllll! Harness the power of the most intuitive interaction and make your websites come alive!

[>>> DEMO <<<](https://alexfox.dev/laxxx/)

![](https://i.imgur.com/Jbkna80.gif) 

[>>> MARIO DEMO <<<](https://alexfox.dev/laxxx/sprite.html)

![](https://i.imgur.com/GGQFucZ.gif)

## Getting Started

### NPM Setup

```bash
npm install lax.js
```
```js
import lax from 'lax.js'
```

### Basic Browser Setup
1) Add lax.js to your html

```html
<script src="lib/lax.min.js" >
```

2) Initialize the plugin

```javascript
window.onload = function() {
	lax.setup() // init

	const updateLax = () => {
		lax.update(window.scrollY)
		window.requestAnimationFrame(updateLax)
	}

	window.requestAnimationFrame(updateLax)
}
```


3) Add class and attributes to the HTML tags you want to animate e.g.
```html
<p class="lax" data-lax-preset="spin fadeInOut">Look at me goooooo!</p>
```

4) Scroll and enjoy!

### Usage With React, Vue.js & DOM Changes
To increase performance lax.js indexes the list of elements to animate when the page loads. If you're using a library like React or vue.js, it is likely that you are adding elements after the initial `window.onload`. Because of this you will need to call `lax.addElement(domElement)` when you add components to the DOM that you want to animate. 

See below for working examples:
* [react](https://codepen.io/alexfoxy/pen/PLaKaE)
* [vue](https://codepen.io/alexfoxy/pen/ZPRZBq)

You can also call `lax.removeElement(domElement)` when the component unmounts.


## Presets
The easiest way to get started is to use the presets via the `data-lax-preset` attribute. You can chain multiple presets together for e.g. `data-lax-preset="blurOut fadeOut spin"`. Some presets also support an optional strength e.g. `data-lax-preset="blurOut-50"`.

See the list of [Supported Presets](#supported-presets) for details.

## Custom Animations

You can easily create your own effects. Just add an attribute to your HTML tag (see [Supported Attribute Keys](#supported-attribute-keys)) with an array of values. These arrays take the format of `scrollPos val, scrollPos val, ... | option=val` e.g:
```html
<p class="lax" data-lax-opacity="0 1, 100 1, 200 0 | loop=200">
	I start to fade out after the window scrolls 100px
	and then I'm gone by 200px!
</p>
```

By default the `scrollPos` is `window.scrollY` but you can use an element distance from the top of the screen instead. You can either pass in a selector `data-lax-anchor="#bio"` or set it to use itself `data-lax-anchor="self"` (this is the default for all presets) e.g.
```html
<p class="lax" data-lax-opacity="200 1, 100 1, 0 0" data-lax-anchor="self">
	I start to fade out after I'm 100px away from the top of the window
	and then I'm gone by the time I reach the top!
</p>
```

### Special Values

| Key     	| Value           |
| ------------- | ------------- |
| vw       	| window.innerWidth  |
| vh     	| window.innerHeight |
| elw     	| targetElement.clientWidth |
| elh     	| targetElement.clientHeight |

You can use these instead of integer values for the scrollPos  e.g.
```html
<p class="lax" data-lax-opacity="0 1, vh 0">
	I fade out as the page scrolls down and
	I'm gone when the page has scrolled the view port height!
</p>
```

### Calculated Values

You can also use vanilla JS within `( )` for calculations and access to more variables e.g.
```html
<p class="lax" data-lax-opacity="0 1, (document.body.scrollHeight*0.5) 0">
	I fade out as the page scrolls down and
	I'm gone when the page has scrolled 50%
	down the entire page height!
</p>
```

### Options

You can pass options into your custom animations for more control e.g. 

```html
<p class="lax" data-lax-opacity="0 1, 100 0, 200 100 | loop=200 offset=100 speed=2">
	I start at 0 opacity and
	fade in and out every 50px
</p>
```

| Option     	| Effect  | 
| ------------- | -------------	| 
| loop      	| modulus the input scrollY position so the animation will loop every `loop` pixels |
| offset     	| add `offset` to scrollY position so the animation will begin at this point |
| speed     	| multiplies the input scrollY position by `speed` to change the speed of the animation |


## Responsive Design
You can set multiple presets and animations for different screen widths. When setting up lax you need to pass in your screen width breakpoints e.g.
```javascript
lax.setup({
    breakpoints: { small: 0, large: 992 }
})
```
Then you can define presets or transforms per breakpoint. 
```html
<p class="lax" data-lax-preset_small="spin">
	I only spin when the screen is smaller than 992px.
</p>

<p class="lax" data-lax-scale_small="0 1, 500 0" data-lax-scale_large="0 1, 500 2">
	I shrink when the screen is smaller than 992px but grow when the screen is larger 992px.
</p>
```

## Supported Presets

| Preset     	| Default Strength | 
| ------------- | -------------	| 
| linger      	| n/a 		|
| lazy     	| 100 		|
| eager     	| 100 		|
| lazy     	| 100 		|
| slalom     	| 50 		|
| crazy     	| n/a 		|
| spin     	| 360 		|
| spinRev     	| 360 		|
| spinIn     	| 360 		|
| spinOut     	| 360 		|
| blurInOut     | 40 		|
| blurIn     	| 40 		|
| blurOut     	| 40 		|
| fadeInOut     | n/a 		|
| fadeIn     	| n/a 		|
| fadeOut     	| n/a 		|
| driftLeft     | 100 		|
| driftRight    | 100 		|
| leftToRight   | 1 		|
| rightToLeft   | 1 		|
| zoomInOut    	| 0.2 		|
| zoomIn     	| 0.2 		|
| zoomOut     	| 0.2 		|
| swing     	| 30		|
| speedy     	| 30 		|

## Supported Attribute Keys

Transforms

| Transform     | Key           |
| ------------- | ------------- |
| opacity       | data-lax-opacity  |
| translate     | data-lax-translate |
| translateX     | data-lax-translate-x |
| translateY     | data-lax-translate-y |
| scale     | data-lax-scale |
| scaleX     | data-lax-scale-x |
| scaleY     | data-lax-scale-y |
| skew     | data-lax-skew |
| skewX     | data-lax-skew-x |
| skewY     | data-lax-skew-y |
| rotate     | data-lax-rotate |

Filters (note - these may be unperformant on low powered machines)

| Filter     | Key           |
| ------------- | ------------- |
| brightness       | data-lax-brightness  |
| contrast     | data-lax-contrast |
| hue-rotate     | data-lax-hue-rotate |
| blur     | data-lax-blur |
| invert     | data-lax-invert |
| saturate     | data-lax-saturate |
| grayscale     | data-lax-grayscale |

Other

| Filter     | Key           |
| ------------- | ------------- |
| background position | data-lax-bg-pos  |
| background position-x | data-lax-bg-pos-x  |
| background position-y | data-lax-bg-pos-y |

	

## Sprite Sheet Animations
You can create animations using sprite sheets. See a demo [here](https://alexfox.dev/laxxx/sprite.html).

The `data-lax-sprite-data` is required and formated like so `[frameWidth, frameHeight, frameCount, columnCount, scrollStep]`. You can either set the image using CSS or the `data-lax-sprite-image` attribute. e.g.

```html
<div 
	class="lax"
	data-lax-sprite-data="500,500,36,36,10"
	data-lax-sprite-image="./spritesheet.png"
/>
```

You can turn a gif or a video into a sprite sheet with this tool: https://ezgif.com/gif-to-sprite

Note: current implimentation requires the element to be the same size as the frame width & height.

## Custom Presets
To avoid duplicate code you can define your own presets with a list of attributes e.g.
```javascript
lax.addPreset("myCoolPreset", function() {
	return { 
		"data-lax-opacity": "(-vh*0.8) 40, (-vh*0.6) 0",
		"data-lax-rotate": "(-vh*2) 1000, (-vh*0.5) 0" 
	}
})
```
You can then access this preset like this:
```html
<p class="lax" data-lax-preset="myCoolPreset">
	I'm the coolest preset in the world 😎
</p>
```

## Performance Tips

* Avoid nesting lax enabled elements within each other, you'll get better performance using lax with smaller elements in the dom tree.
* Avoid transforms on large elements, e.g. full screen backgrounds.
* By default elements that have opacity 0 aren't updated. You can either manually set up a `data-lax-opacity` to control this yourself or use `data-lax-optimize` which will set the elements opacity to 0 when it goes off -screen. 
* By default `-webkit-backface-visibility: hidden;` is added to your elements style to encourage the browser to render that object as a layer on the GPU and increase performance. To turn this off add `data-lax-use-gpu="false"` to your element.


## Notes

### Screen Rotating & Resizing 
As some values (vh, vw, elh, elw) are calculated on load, when the screen size changes or rotates you might want to recalculate these. E.g.
```
window.addEventListener("resize", function() {
	lax.populateElements()
});
```
Be warned, on mobile, a resize event is fired when you scroll and the toolbar is hidden so you might want to check if the width or orientation has changed.

### Scroll Wheels
Scroll wheels only increment the scroll position in steps which can cause the animations to look janky. You can use the SmoothScroll (http://www.smoothscroll.net/) plugin to smooth this out, however there maybe performance implications that need investigating.

### Merging Existing Styles
Only inline styles for transforms and filters will be merged in to the animation. Transforms and filters derived from CSS will be overwritten.

## To Do / Ideas
* Implement a tween for scroll wheels to remove reliance on smoothscroll
* A way to add weight/momentum to moving objecs
* More cool demos
* More concise read me for react / vue.js
* ~~Support for sprite sheet animations~~



changelog
- Added responsive stuff
- Changed demo to use window.animationFrame
- breaking change - now have to add lax class to elements
- Now merges in elements original transform and filter // inline 
- Can add options into e.g. | loop=360 offset=1000 speed=2
- performance improvements

todo
- update change log
	- usage with vue/react

- add link to my personal site



src/lax.js
//
// lax v0.0.1 (Alex Fox)
// Simple & light weight vanilla javascript plugin to create beautiful animations things when you scrolllll!!
//
// Licensed under the terms of the MIT license.
//
// You may use it in your theme if you credit me. 
// It is also free to use on any individual website.
//
// Exception:
// The only restriction is to not publish any  
// extension for browsers or native application
// without getting a written permission first.
//

(() => {
  const lax = (() => {
    const lax = {
      elements: []
    }

    let lastY = 0;
    let currentBreakpoint = 'default'
    const breakpointsSeparator = "_"

    const transformFns = {
      "data-lax-opacity": (style, v) => { style.opacity = v },
      "data-lax-translate": (style, v) => { style.transform += ` translate(${v}px, ${v}px)` },
      "data-lax-translate-x": (style, v) => { style.transform += ` translateX(${v}px)` },
      "data-lax-translate-y": (style, v) => { style.transform += ` translateY(${v}px)` },
      "data-lax-scale": (style, v) => { style.transform += ` scale(${v})` },
      "data-lax-scale-x": (style, v) => { style.transform += ` scaleX(${v})` },
      "data-lax-scale-y": (style, v) => { style.transform += ` scaleY(${v})` },
      "data-lax-skew": (style, v) => { style.transform += ` skew(${v}deg, ${v}deg)` },
      "data-lax-skew-x": (style, v) => { style.transform += ` skewX(${v}deg)` },
      "data-lax-skew-y": (style, v) => { style.transform += ` skewY(${v}deg)` },
      "data-lax-rotate": (style, v) => { style.transform += ` rotate(${v}deg)` },
      "data-lax-brightness": (style, v) => { style.filter += ` brightness(${v}%)` },
      "data-lax-contrast": (style, v) => { style.filter += ` contrast(${v}%)` },
      "data-lax-hue-rotate": (style, v) => { style.filter += ` hue-rotate(${v}deg)` },
      "data-lax-blur": (style, v) => { style.filter += ` blur(${v}px)` },
      "data-lax-invert": (style, v) => { style.filter += `  invert(${v}%)` },
      "data-lax-saturate": (style, v) => { style.filter += `  saturate(${v}%)` },
      "data-lax-grayscale": (style, v) => { style.filter += `  grayscale(${v}%)` },
      "data-lax-bg-pos": (style, v) => { style.backgroundPosition = `${v}px ${v}px` },
      "data-lax-bg-pos-x": (style, v) => { style.backgroundPositionX = `${v}px` },
      "data-lax-bg-pos-y": (style, v) => { style.backgroundPositionY = `${v}px` }
    }

    let crazy = ""

    for(let i=0;i<20;i++) {
      crazy += " " + i*5 + " " + (i*47)%360 + ", "
    }

    lax.presets = {
      linger: (v) => {
        return { "data-lax-translate-y": `(vh*0.7) 0, 0 200, -500 0` }
      },
      lazy: (v=100) => {
        return { "data-lax-translate-y": `(vh) 0, (-elh) ${v}` }
      },
      eager: (v=100) => {
        return { "data-lax-translate-y": `(vh) 0, (-elh) -${v}` }
      },
      slalom: (v=50) => {
        return { "data-lax-translate-x": `vh ${v}, (vh*0.8) ${-v}, (vh*0.6) ${v}, (vh*0.4) ${-v}, (vh*0.2) ${v}, (vh*0) ${-v}, (-elh) ${v}` }
      },
      crazy: (v) => {
        return { "data-lax-hue-rotate": `${crazy} | loop=20` }
      },
      spin: (v=360) => {
        return { "data-lax-rotate": `(vh) 0, (-elh) ${v}` }
      },
      spinRev: (v=360) => {
        return { "data-lax-rotate": `(vh) 0, (-elh) ${-v}` }
      },
      spinIn: (v=360) => {
        return { "data-lax-rotate": `vh ${v}, (vh*0.5) 0` }
      },
      spinOut: (v=360) => {
        return { "data-lax-rotate": `(vh*0.5) 0, -elh ${v}` }
      },
      blurInOut: (v=40) => {
        return { "data-lax-blur": `(vh) ${v}, (vh*0.8) 0, (vh*0.2) 0, 0 ${v}` }
      },
      blurIn: (v=40) => {
        return { "data-lax-blur": `(vh) ${v}, (vh*0.7) 0` }
      },
      blurOut: (v=40) => {
        return { "data-lax-blur": `(vh*0.3) 0, 0 ${v}` }
      },
      fadeInOut: () => {
        return { "data-lax-opacity": "(vh) 0, (vh*0.8) 1, (vh*0.2) 1, 0 0" }
      },
      fadeIn: () => {
        return { "data-lax-opacity": "(vh) 0, (vh*0.7) 1" }
      },
      fadeOut: () => {
        return { "data-lax-opacity": "(vh*0.3) 1, 0 0" }
      },
      driftLeft: (v=100) => {
        return { "data-lax-translate-x": `vh ${v}, -elh ${-v}` }
      },
      driftRight: (v=100) => {
        return { "data-lax-translate-x": `vh ${-v}, -elh ${v}` }
      },
      leftToRight: (v=1) => {
        return { "data-lax-translate-x": `vh 0, -elh (vw*${v})` }
      },
      rightToLeft: (v=1) => {
        return { "data-lax-translate-x": `vh 0, -elh (vw*${-v})` }
      },
      zoomInOut: (v=0.2) => {
        return { "data-lax-scale": `(vh) ${v}, (vh*0.8) 1, (vh*0.2) 1, -elh ${v}` }
      },
      zoomIn: (v=0.2) => {
        return { "data-lax-scale": `(vh) ${v}, (vh*0.7) 1` }
      },
      zoomOut: (v=0.2) => {
        return { "data-lax-scale": `(vh*0.3) 1, -elh ${v}` }
      },
      speedy: (v=30) => {
        return { "data-lax-skew-x": `(vh) ${v}, -elh ${-v}` }
      },
      swing: (v=30) => {
        return { "data-lax-skew-y": `(vh) ${v}, -elh ${-v}` }
      }
    }

    lax.addPreset = (p, o) => {
      lax.presets[p] = o
    }

    function intrp(t, v) {
      let i = 0

      while(t[i][0] <= v && t[i+1] !== undefined) {
        i+=1
      }

      const x = t[i][0]
      const prevX = t[i-1] === undefined ? x : t[i-1][0]

      const y = t[i][1]
      const prevY = t[i-1] === undefined ? y : t[i-1][1]

      const xPoint = Math.min(Math.max((v-prevX)/(x-prevX),0),1)
      const yPoint = (xPoint*(y-prevY)) + prevY

      return yPoint
    }

    function fnOrVal(s) {
      if(s[0] === "(") return eval(s)
      else return parseFloat(s)
    }

    lax.setup = (o={}) => {
      lax.breakpoints = o.breakpoints || {}
      
      lax.selector = o.selector || '.lax'
      lax.populateElements()
    }

    lax.removeElement = (el) => {
      const i = lax.elements.findIndex(o => o.el = el)
      if(i > -1) {
        lax.elements.splice(i, 1)
      }
    }

    lax.addElement = (el) => {
      const o = {
        el: el,
        originalStyle: {
          transform: el.style.transform,
          filter: el.style.filter
        },
        transforms: {}
      }

      //find presets in data attributes
      const presets = []
      for(let i=0; i<el.attributes.length; i++) {
          const a = el.attributes[i]
          if(a.name.indexOf("data-lax-preset") > -1)  {
            presets.push(a);
          }
      }

      //unwrap presets into transformations
      for(let i=0; i<presets.length; i++) {
          const a = presets[i]
          const b = a.name.split(breakpointsSeparator)
          const breakpoint = b[1] ? `${breakpointsSeparator}${b[1]}` : ''
          a.value.split(" ").forEach((p) => {
              const bits = p.split("-")
              const fn = lax.presets[bits[0]]
              if(!fn) {
                  console.log(`lax error: preset ${bits[0]} is not defined`)
              } else {
                  const d = fn(bits[1])
                  for(let k in d) {
                      el.setAttribute(`${k}${breakpoint}`, d[k])
                  }
              }
          })

          el.setAttribute("data-lax-anchor", "self")
          el.attributes.removeNamedItem(a.name)

      }

      // use gpu
      const useGpu = !(el.attributes["data-lax-use-gpu"] && el.attributes["data-lax-use-gpu"].value === 'false')
      if(useGpu) {
        el.style["backface-visibility"] = "hidden"
        el.style["-webkit-backface-visibility"] = "hidden"
      }
      if(el.attributes["data-lax-use-gpu"]) el.attributes.removeNamedItem("data-lax-use-gpu")

      // optmise off screen
      o.optimise = false 
      if(el.attributes["data-lax-optimize"] && el.attributes["data-lax-optimize"].value === 'true') {
        o.optimise = true
        const bounds = el.getBoundingClientRect()
        el.setAttribute("data-lax-opacity", `${-bounds.height-1} 0, ${-bounds.height} 1, ${window.innerHeight} 1, ${window.innerHeight+1} 0`)
        el.attributes.removeNamedItem("data-lax-optimize")
      }

      // build transform list
      for(let i=0; i<el.attributes.length; i++) {
        const a = el.attributes[i]
        if(a.name.indexOf("data-lax") < 0) continue

        const b = a.name.split(breakpointsSeparator)
        const bits = b[0].split("-")
        const breakpoint = b[1] || "default"
        if(bits[1] === "lax") {
          if(a.name === "data-lax-anchor") {
            o["data-lax-anchor"] = a.value === "self" ? el : document.querySelector(a.value)
            const rect = o["data-lax-anchor"].getBoundingClientRect()
            o.anchorTop = Math.floor(rect.top) + window.scrollY
          } else {
            const tString = a.value
              .replace(/vw/g, window.innerWidth)
              .replace(/vh/g, window.innerHeight)
              .replace(/elh/g, el.clientHeight)
              .replace(/elw/g, el.clientWidth)
              .replace(/\s+/g," ")

            const [arrString, optionString] = tString.split("|")
            const options = {}

            if(optionString) {
              optionString.split(" ").forEach((o) => {
                const [key, val] = o.split("=")
                options[key] = key && fnOrVal(val)
              }) 
            }

            const name = b[0]
            const valueMap = arrString.split(",").map((x) => { 
                return x.trim().split(" ").map(fnOrVal)
              }).sort((a,b) => {
                return a[0] - b[0]  
              })

            if(!o.transforms[name]) {
              o.transforms[name] = {}
            }
            
            o.transforms[name][breakpoint] = { valueMap, options } 
          }
        }
      }

      // sprite sheet animation
      const spriteData = el.attributes["data-lax-sprite-data"] && el.attributes["data-lax-sprite-data"].value
      if(spriteData) {
        o.spriteData = spriteData.split(",").map(x => parseInt(x))
        el.style.height = o.spriteData[1] + "px"
        el.style.width = o.spriteData[0] + "px"

        const spriteImage = el.attributes["data-lax-sprite-image"] && el.attributes["data-lax-sprite-image"].value
        if(spriteImage) {
          el.style.backgroundImage = `url(${spriteImage})`
        }
      }

      lax.elements.push(o)
      lax.updateElement(o)
    }

    lax.populateElements = () => {
      lax.elements = []

      document.querySelectorAll(lax.selector).forEach(lax.addElement)
      currentBreakpoint = lax.getCurrentBreakPoint()
    }

    lax.getCurrentBreakPoint = () => {
      let b = 'default'
      const w = window.innerWidth

      for(let i in lax.breakpoints) {
        const px = parseFloat(lax.breakpoints[i])
        if(px <= w) {
          b = i
        } else {
          break
        }
      }

      return b
    }

    lax.updateElement = (o) => {
      const { originalStyle, anchorTop, transforms, spriteData, el } = o

      let r = anchorTop ? anchorTop-lastY : lastY

      const style = {
        transform: originalStyle.transform,
        filter: originalStyle.filter
      }

      for(let i in transforms) {
        const transformData = transforms[i][currentBreakpoint] || transforms[i]["default"]

        if(!transformData) {
          // console.log(`lax error: there is no setting for key ${i} and screen size ${currentBreakpoint}. Try adding a default value!`)
          continue
        }

        let _r = r
        if(transformData.options.offset) _r = _r+transformData.options.offset
        if(transformData.options.speed) _r = _r*transformData.options.speed
        if(transformData.options.loop) _r = _r%transformData.options.loop

        const t = transformFns[i]
        const v = intrp(transformData.valueMap, _r)

        if(!t) {
          // console.info(`lax: error ${i} is not supported`)
          continue
        }

        t(style, v)
      }

      if(spriteData) {
        const [frameW,frameH,numFrames,cols,scrollStep] = spriteData
        const frameNo = Math.floor(lastY/scrollStep) % numFrames
        const framePosX = frameNo%cols
        const framePosY = Math.floor(frameNo/cols)
        style.backgroundPosition = `-${framePosX*frameW}px -${framePosY*frameH}px`
      }

      if(style.opacity === 0) { // if opacity 0 don't update
        el.style.opacity = 0 
      } else {
        for(let k in style) {
          el.style[k] = style[k]
        }
      }
    }

    lax.update = (y) => {
      if(lastY === y) return
      lastY = y
      lax.elements.forEach(lax.updateElement)
    }

    return lax;
  })();

  if (typeof module !== 'undefined' && typeof module.exports !== 'undefined')
    module.exports = lax;
  else
    window.lax = lax;
})();
